### PR TITLE
Refresh testing guide

### PR DESCRIPTION
**Pull request by hcindyl**
Update the available testing targets and their support

* Provide a table of targets with unit --> e2e complexity
* Re-order compiler tests and runtime tests

### CODE REVIEW

File: docs/developers/developing_iree/testing_guide.md [hunk 1]
```diff
@@ -3,6 +3,96 @@
...
+|                 | iree_hal_cts_test_suite                         | CMake        | x86_64/arm64-v8a/riscv32/riscv64 |
+| Core E2E tests  | iree_check_test                                 | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_check_test_suite                           | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_check_single_backend_test_suite            | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_trace_runner_test                          | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_single_backend_generated_trace_runner_test | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_generated_trace_runner_test                | Bazel/CMake  | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_run_module_test                            | CMake        | x86_64/arm64-v8a/riscv32/riscv64 |
+|                 | iree_static_linker_test                         | CMake        | x86_64/arm64-v8a/riscv32/riscv64 |
```
Comment (ScottTodd) [inline docs/developers/developing_iree/testing_guide.md:20]: I'd just write "Host/Device" or something more generic for supported platforms (these also run on Emscripten, for example)

File: docs/developers/developing_iree/testing_guide.md [hunk 2]
```diff
@@ -3,6 +3,96 @@
 Like the IREE project in general, IREE tests are divided into a few different
 components and use different tooling depending on the needs of that component.
 
+| Test type       | Test                                            | Build system | Supported platforms              |
+|:--------------  | :-----------------                              | -----------  | -------------                    |
+| Compiler tests  | iree_lit_test                                   | Bazel/CMake  | Host                             |
+|                 | iree_lit_test_suite                             | Bazel/CMake  | Host                             |
```
Comment (ScottTodd) [inline docs/developers/developing_iree/testing_guide.md:9]: You can omit the `*_test_suite` "test types" where there is a corresponding `*_test` from this table IMO - a suite is just a grouping of tests

**Review by ScottTodd — APPROVED** [review]
Thanks!

File: docs/developers/developing_iree/testing_guide.md (hunk 1 above)
Comment (hcindyl) [reply docs/developers/developing_iree/testing_guide.md]: Done.

File: docs/developers/developing_iree/testing_guide.md (hunk 2 above)
Comment (hcindyl) [reply docs/developers/developing_iree/testing_guide.md]: Done. I leave `iree_hal_cts_test_suite` in the table, as it does not have a single test. I also put some note of `*_test_suite` after the table.

File: docs/developers/developing_iree/testing_guide.md [hunk 3]
```diff
@@ -3,6 +3,95 @@
...
+|:--------------  | :-----------------                              | -----------  | -------------       |
+| Compiler tests  | iree_lit_test                                   | Bazel/CMake  | Host                |
+| Runtime tests   | iree_cc_test                                    | Bazel/CMake  | Host/Device         |
+|                 | iree_native_test                                | Bazel/CMake  | Host/Device         |
+|                 | iree_hal_cts_test_suite                         | CMake        | Host/Device         |
+| Core E2E tests  | iree_check_test                                 | Bazel/CMake  | Host/Device         |
+|                 | iree_trace_runner_test                          | Bazel/CMake  | Host/Device         |
+|                 | iree_single_backend_generated_trace_runner_test | Bazel/CMake  | Host/Device         |
+|                 | iree_generated_trace_runner_test                | Bazel/CMake  | Host/Device         |
+|                 | iree_static_linker_test                         | CMake        | Host                |
```
Comment (hcindyl) [inline docs/developers/developing_iree/testing_guide.md:16]: Remove `iree_run_module_test` and the `Host/Device` support for `iree_static_linker_test`. I will update the table after the corresponding test functions actually landed.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**